gnosis/cowswap
# [presign] Enhanced safe transactions

# Summary

This PR adds some context for Gnosis Safe transactions.

* It will show when the transaction is a Gnosis Safe, it will show also the nonce so you can identify it within the safe
* While the tx has some pending signers, it will show how many signatures are left
* It includes a smart link 

The smart link is basically, that before transactions would take you always to Etherscan. Now, they will take you to:
* Etherscan for non safe tx
* Gnosis Safe Web for safe tx that are not mined
* Etherscan for safe tx that are mined

When the transaction is mined for a  Gnosis Safe tx, it will additionally show a link to Gnosis Safe Web, so the user will have the option to click in the tx to see it in Etherscan, or click in the link and see it in Gnosis Safe.

This PR applies only to transactions, not to orders.

Example of waiting for signatures safe tx:
![image](https://user-images.githubusercontent.com/2352112/135160669-d9007848-e2c6-450d-b9e2-49dbc6b1f121.png)

Example of mined gnosis safe tx:
<img width="687" alt="Screenshot at Sep 28 20-33-30" src="https://user-images.githubusercontent.com/2352112/135160704-bc49cb6a-740a-4cb4-b7b0-b7efdb6f5417.png">


Additionally, this PR does also a big refactor in the transaction component to break it into more manageable pieces.

## Not included
Please, don't focus the review on the design. @biocom will probably want to reiterate. This is basically setting the grounds so we can shape it in the future as we please.

Doesn't include any change for orders. If we do something for orders workflow, it will be in a future PR.

Doesn't solve two issues (they were already in the base branch) and will be address separatelly:
* Unwrap doesn't work: https://github.com/gnosis/cowswap/issues/1469
* Reject using Wallet Connect is not working


## Notes to the reviewer


This is a rather big PR in terms of changes, but I think with a few hints, it should be easier to review.

### Partition into components
The transaction was a monolithic component that was handling the orders and transactions that we represent in the side panel.




This PR needed to add a few elements to this representation to handle Gnosis Safe transactions, there fore, it was going to increase it's complexity. 

I took this chance to break the component into smaller parts. Basically 3 main parts:
![image](https://user-images.githubusercontent.com/2352112/135158949-ce2ef516-715b-46e5-966a-c345ff1a49c7.png)

These tree parts are easy to spot in the way smaller component that render activities:
![image](https://user-images.githubusercontent.com/2352112/135159117-262efc95-e81f-42bc-8a82-d0c5b7383d03.png)

This partition was done also into different files, so now all components are in a manageable size file:
![image](https://user-images.githubusercontent.com/2352112/135158401-774b37c7-a2ea-4a16-820c-945304f00dd3.png)

### ActivityDerivedState
This is a object that is constructed with some data related to the activity. Is derived data meant to simplify tendering all the components.

The creation of this object is done using an aux function `getActivityDerivedState` and it's result  is memoized.


# To Test

1. Create transactions (approve/wrap) using EOA
2. Create transactions (approve/wrap) using Gnosis Sage

For both cases, observe all the states the transaction transitions. Test the links.

Known issues that won't be solved in this PR: 
* Unwrap doesn't work: https://github.com/gnosis/cowswap/issues/1469
* Reject using Wallet Connect is not working
* Also be aware of the "Not Included: section
